src/NVMLDevicePool.cpp

```diff
@@ -0,0 +1,367 @@
+
+        return (uint64_t)result;
+    }
+
+    uint64_t NVMLDevicePoolImp::energy(int accel_idx) const
+    {
+        check_accel_range(accel_idx);
+        unsigned long long result;
+        nvmlReturn_t nvml_result;
+
+        nvml_result =  nvmlDeviceGetTotalEnergyConsumption(m_nvml_device.at(accel_idx), &result);
+        check_nvml_result(nvml_result, GEOPM_ERROR_RUNTIME, "NVMLDevicePool::" + std::string(__func__) +
+                          ": NVML failed to get energy for accelerator " +
+                          std::to_string(accel_idx) + ".");
+
+        return (uint64_t)result;
+    }
+
+    uint64_t NVMLDevicePoolImp::performance_state(int accel_idx) const
+    {
+        check_accel_range(accel_idx);
+        nvmlPstates_t result;
+        nvmlReturn_t nvml_result;
+
+        nvml_result = nvmlDeviceGetPerformanceState(m_nvml_device.at(accel_idx), &result);
+        check_nvml_result(nvml_result, GEOPM_ERROR_RUNTIME, "NVMLDevicePool::" + std::string(__func__) +
+                          ": NVML failed to get performance state for accelerator " +
+                          std::to_string(accel_idx) + ".");
+
+        return (uint64_t)result;
+    }
+
+    uint64_t NVMLDevicePoolImp::throughput_rx_pcie(int accel_idx) const
+    {
+        check_accel_range(accel_idx);
+        unsigned int result;
+        nvmlReturn_t nvml_result;
+
+        nvml_result = nvmlDeviceGetPcieThroughput(m_nvml_device.at(accel_idx), NVML_PCIE_UTIL_RX_BYTES, &result);
+        check_nvml_result(nvml_result, GEOPM_ERROR_RUNTIME, "NVMLDevicePool::" + std::string(__func__) +
+                          ": NVML failed to get PCIE received throughput rate for accelerator " +
+                          std::to_string(accel_idx) + ".");
+
+        return (uint64_t)result;
+    }
+
+    uint64_t NVMLDevicePoolImp::throughput_tx_pcie(int accel_idx) const
+    {
+        check_accel_range(accel_idx);
+        unsigned int result;
+        nvmlReturn_t nvml_result;
+
+        nvml_result = nvmlDeviceGetPcieThroughput(m_nvml_device.at(accel_idx), NVML_PCIE_UTIL_TX_BYTES, &result);
+        check_nvml_result(nvml_result, GEOPM_ERROR_RUNTIME, "NVMLDevicePool::" + std::string(__func__) +
+                          ": NVML failed to get PCIE transmitted throughput rate for accelerator " +
+                          std::to_string(accel_idx) + ".");
+
+        return (uint64_t)result;
+    }
+
+    double NVMLDevicePoolImp::utilization_mem(int accel_idx) const
+    {
+        check_accel_range(accel_idx);
+        nvmlUtilization_t result;
+        nvmlReturn_t nvml_result;
+
+        nvml_result =  nvmlDeviceGetUtilizationRates(m_nvml_device.at(accel_idx), &result);
+        check_nvml_result(nvml_result, GEOPM_ERROR_RUNTIME, "NVMLDevicePool::" + std::string(__func__) +
+                          ": NVML failed to get memory utilization for accelerator " +
+                          std::to_string(accel_idx) + ".");
+
+        return (double)result.memory;
```

Comment:
This is in bytes?

Reply:
https://docs.nvidia.com/deploy/nvml-api/structnvmlUtilization__t.html#structnvmlUtilization__t
The result here is a struct that contain GPU and Memory Utilization as an unsigned int with values from 0-100.  

This (and the gpu utilizaiton) should probably be returned as uint64_t, with casting to double to convert the scale to 0.0-1.0 should happen in the IOGroup.